### PR TITLE
Respect user settings in php.ini if they are big enough

### LINKED ISSUE #32209: [Bug]:  Imposible to set max_execution_time and max_input_time in php.ini
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

When configuring max_execution_time and max_input_time in php.ini the new settings will not take and settings/admin/serverinfo will forever show 3600 no matter what you set it to. This is not the case with other settings such as max_file_size.

I am pretty sure that the problem is these two lines of code where these two variables are hard coded to 3600:

https://github.com/nextcloud/server/blob/135bdb3d5830672214149fa0b75fadd29b20b844/lib/base.php#L634-L635

Changing the hard coded value will make the web interface update to the value you set, indicating that indeed Nextcloud is using the hard coded value rather than the value the administrator has configured in php.ini.  

### Steps to reproduce

1. Update max_execution_time and max_input_time in php.ini

### Expected behavior

Settings are honored by Nextcloud, and reported in settings/admin/serverinfo 

### Installation method

Other

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.0

### Web server

Apache (supported)

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

_No response_

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [x] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [X] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "memcache.local": "\\OC\\Memcache\\APCu",
        "appstoreenabled": false,
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "password": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "forcessl": true,
        "overwriteprotocol": "https",
        "objectstore": {
            "class": "\\OC\\Files\\ObjectStore\\S3",
            "arguments": {
                "bucket": "***REMOVED SENSITIVE VALUE***",
                "key": "***REMOVED SENSITIVE VALUE***",
                "secret": "***REMOVED SENSITIVE VALUE***",
                "region": "us-east-1",
                "hostname": "***REMOVED SENSITIVE VALUE***",
                "port": "",
                "objectPrefix": "urn:oid:",
                "autocreate": false,
                "use_ssl": true,
                "use_path_style": true,
                "legacy_auth": false
            }
        },
        "csrf.disabled": true,
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***",
            "localhost",
            "***REMOVED SENSITIVE VALUE***",
            "***REMOVED SENSITIVE VALUE***",
            "***REMOVED SENSITIVE VALUE***"
        ],
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "config_is_read_only": true,
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "dbport": "3306",
        "dbtableprefix": "oc_",
        "dbtype": "mysql",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "gs.enabled": "true",
        "gs.federation": "global",
        "gs.trustedHosts": [
            "***REMOVED SENSITIVE VALUE***"
        ],
        "gss.discovery.manual.mapping.file": "\/var\/www\/html\/mappingfile.json",
        "gss.discovery.manual.mapping.parameter": "urn:oid:1.3.6.1.4.1.5923.1.1.1.6",
        "gss.discovery.manual.mapping.regex": true,
        "gss.jwt.key": "***REMOVED SENSITIVE VALUE***",
        "gss.master.admin": [
            "admin",
            "***REMOVED SENSITIVE VALUE***",
            "***REMOVED SENSITIVE VALUE***",
            "***REMOVED SENSITIVE VALUE***",
            "***REMOVED SENSITIVE VALUE***"
        ],
        "gss.master.url": "***REMOVED SENSITIVE VALUE***",
        "gss.mode": "slave",
        "gss.user.discovery.module": "\\OCA\\GlobalSiteSelector\\UserDiscoveryModules\\ManualUserMapping",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "log_type": "file",
        "loglevel": 0,
        "lookup_server": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_sendmailmode": "smtp",
        "mail_smtpauth": 1,
        "mail_smtpauthtype": "LOGIN",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "mail_smtpsecure": "tls",
        "mail_template_class": "OCA\\DriveEmailTemplate\\EMailTemplate",
        "mysql.utf8mb4": true,
        "overwritehost": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
        "templatedirectory": "",
        "skeletondirectory": "",
        "twofactor_enforced": "true",
        "twofactor_enforced_groups": [
            "admin",
            "***REMOVED SENSITIVE VALUE***"
        ],
        "twofactor_enforced_excluded_groups": [],
        "version": "21.0.9.2",
        "app_install_overwrite": [
            "globalsiteselector"
        ]
    }
}
```

### List of activated Apps

```shell
Enabled:
  - accessibility: 1.7.0
  - activity: 2.14.3
  - admin_audit: 1.11.0
  - cloud_federation_api: 1.4.0
  - comments: 1.11.0
  - contactsinteraction: 1.2.0
  - dashboard: 7.1.0
  - dav: 1.17.1
  - drive_email_template: 0.1
  - federatedfilesharing: 1.11.0
  - federation: 1.11.0
  - files: 1.16.0
  - files_external: 1.12.1
  - files_pdfviewer: 2.1.1
  - files_rightclick: 1.0.0
  - files_sharing: 1.13.1
  - files_trashbin: 1.11.0
  - files_versions: 1.14.0
  - files_videoplayer: 1.10.0
  - firstrunwizard: 2.10.0
  - logreader: 2.6.0
  - lookup_server_connector: 1.9.0
  - nextcloud_announcements: 1.10.0
  - notifications: 2.9.0
  - oauth2: 1.9.0
  - password_policy: 1.11.0
  - photos: 1.3.0
  - privacy: 1.5.0
  - provisioning_api: 1.11.0
  - recommendations: 1.0.0
  - richdocuments: 4.2.4
  - serverinfo: 1.11.0
  - settings: 1.3.0
  - sharebymail: 1.11.0
  - support: 1.4.0
  - survey_client: 1.9.0
  - systemtags: 1.11.0
  - text: 3.2.0
  - theming: 1.12.0
  - twofactor_backupcodes: 1.10.0
  - twofactor_totp: 6.1.0
  - twofactor_u2f: 6.2.0
  - updatenotification: 1.11.0
  - user_status: 1.1.1
  - viewer: 1.5.0
  - weather_status: 1.1.0
  - workflowengine: 2.3.1
Disabled:
  - encryption
  - files_markdown
  - files_texteditor
  - globalsiteselector: 2.1.0
  - loginpagebutton
  - onlyoffice: 7.2.0
  - twofactor_admin
  - user_ldap
  - user_saml
```

### Nextcloud Signing status

```shell
N/A
```

### Nextcloud Logs

```shell
N/A
```

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by mickenordin**
In the admin guide:
* https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/big_file_upload_configuration.html

Fix https://github.com/nextcloud/server/issues/32209

it is mentioned that you can tweek:
* max_input_time
* max_execution_time

in order to enable larger file uploads. However, the current codebase
will hard code these values to one hour, no matter what the user sets in
php.ini.

This patch will allow the user to set these settings in php.ini and they
will be respected, if and only if, they are set to something bigger than
3600 seconds.

### CODE REVIEW

**Comment by mickenordin** [comment]
This fixes #32209

**Comment by szaimen** [comment]
Thanks for your PR! can you please solve DCO (see https://github.com/nextcloud/server/pull/32216/checks?check_run_id=6223623697)? Thank you!

**Comment by mickenordin** [comment]
> Thanks for your PR! can you please solve DCO (see https://github.com/nextcloud/server/pull/32216/checks?check_run_id=6223623697)? Thank you!

Sure, fixed.

File: lib/base.php [hunk 1]
```diff
@@ -623,16 +623,32 @@ public static function init() {
...
+		//Let´s try to overwrite some defaults if they are smaller than 1 hour
+		// One hour is 3600 seconds
+		$time_limit = 3600;
+
+		$max_execution_time_from_ini = @ini_get('max_execution_time');
+		$biggest_max_execution_time = $time_limit;
+		if (isset($max_execution_time_from_ini)) {
+			$biggest_max_execution_time = max($time_limit, intval($max_execution_time_from_ini));
+		}
+		@ini_set('max_execution_time', strval($biggest_max_execution_time));
```
Comment (artonge) [inline lib/base.php:638]: Seems a bit complicated, would a simpler version work?

```suggestion
	if (intval(@ini_get('max_execution_time')?? 0) < 3600) {
		@ini_set('max_execution_time', strval(3600));
	}
```

Comment (solracsf) [reply lib/base.php:638]: Well, i don't really appreciate @ in the code...but it's my own opinion.
Better to check the function exists before calling it instead.

Comment (mickenordin) [reply lib/base.php]: I have implemented the simpler solution now :)

**Comment by kesselb** [comment]
Thanks for sending a pull request :+1: 

The above code should work and allow an administrator to pick a higher value. 

I think the actual issue here is that we configure `max_execution_time` and `max_input_time` at all. That code should go and we introduce a setup check to inform the administrator that an update for php.ini / user.ini / .htaccess is required to configure good limits. This auto configuration approach might be nice but it's always a pain if one needs a higher limit. My take here is that Nextcloud should not try to auto configure the environment but tell the administrator what's necessary. Less magic => less pain ;)

Also keep in mind that almost **every** request is executing this code. It's already bad to have the ini_set calls (with suppress operator) here. If we merge this patch we check on each time if there is a higher limit and if not run the ini_set call. I don't think that removing those calls will improve the performance yet it's still bad.

**Comment by CarlSchwan** [comment]
> Thanks for sending a pull request +1
> 
> The above code should work and allow an administrator to pick a higher value.
> 
> I think the actual issue here is that we configure `max_execution_time` and `max_input_time` at all. That code should go and we introduce a setup check to inform the administrator that an update for php.ini / user.ini / .htaccess is required to configure good limits. This auto configuration approach might be nice but it's always a pain if one needs a higher limit. My take here is that Nextcloud should not try to auto configure the environment but tell the administrator what's necessary. Less magic => less pain ;)
> 
> Also keep in mind that almost **every** request is executing this code. It's already bad to have the ini_set calls (with suppress operator) here. If we merge this patch we check on each time if there is a higher limit and if not run the ini_set call. I don't think that removing those calls will improve the performance yet it's still bad.

For shared hosting providers, it's often not possible to modify the ini files, this is probably the reason why this is done like this currently.

**Comment by solracsf** [comment]
> For shared hosting providers, it's often not possible to modify the ini files, this is probably the reason why this is done like this currently.

For shared hosting, most of the time, both are disabled (ini files and ini_set())

**Comment by mickenordin** [comment]
> Thanks for sending a pull request 👍
> 
> The above code should work and allow an administrator to pick a higher value.
> 
> I think the actual issue here is that we configure `max_execution_time` and `max_input_time` at all. That code should go and we introduce a setup check to inform the administrator that an update for php.ini / user.ini / .htaccess is required to configure good limits. This auto configuration approach might be nice but it's always a pain if one needs a higher limit. My take here is that Nextcloud should not try to auto configure the environment but tell the administrator what's necessary. Less magic => less pain ;)
> 
> Also keep in mind that almost **every** request is executing this code. It's already bad to have the ini_set calls (with suppress operator) here. If we merge this patch we check on each time if there is a higher limit and if not run the ini_set call. I don't think that removing those calls will improve the performance yet it's still bad.

I am not married to either solution, I just want to be able to change the setting and not have it hard coded, which I think is the worst solution of them all.

This patch is proposed because it is backwards compatible with the old solution. I don't know if removing the hard coded value altogether will break stuff for a lot of people on shared hosting or for people who did not have to think about this untill now. I also don't think there is a lot of performance penalty with doing it this way so it shouldn't be a problem, but if you guys would rather remove the hard coded value that is fine too for me. :)

**Comment by solracsf** [comment]
Sure, this is a "problem" of all `ini_set()` calls at `lib/base.php` ;)

File: lib/base.php [hunk 2]
```diff
@@ -623,16 +623,23 @@ public static function init() {
...
 		}
 
+
 		//try to configure php to enable big file uploads.
-		//this doesn´t work always depending on the web server and php configuration.
-		//Let´s try to overwrite some defaults anyway
+		//this doesn´t work always depending on the webserver and php configuration.
+		//Let´s try to overwrite some defaults if they are smaller than 1 hour
+
+		if (intval(@ini_get('max_execution_time')?? 0) < 3600) {
```
Comment (artonge) [inline lib/base.php:631]:
```suggestion
		if (intval(@ini_get('max_execution_time') ?? 0) < 3600) {
```

File: lib/base.php [hunk 3]
```diff
@@ -623,16 +623,23 @@ public static function init() {
...
+			@ini_set('max_execution_time', strval(3600));
+		}
+
+		if (intval(@ini_get('max_input_time')?? 0) < 3600) {
```
Comment (artonge) [inline lib/base.php:635]:
```suggestion
		if (intval(@ini_get('max_input_time') ?? 0) < 3600) {
```

File: lib/base.php [hunk 4]
```diff
@@ -623,16 +623,23 @@ public static function init() {
...
+			@ini_set('max_input_time', strval(3600));
+		}
 
-		//try to set the maximum execution time to 60min
+		//try to set the maximum execution time to the largest time limit we have
 		if (strpos(@ini_get('disable_functions'), 'set_time_limit') === false) {
-			@set_time_limit(3600);
+			@set_time_limit(strval(max(intval(@ini_get('max_execution_time')),intval(@ini_get('max_input_time')))));
```
Comment (artonge) [inline lib/base.php:641]:
```suggestion
			@set_time_limit(max(intval(@ini_get('max_execution_time')), intval(@ini_get('max_input_time'))));
```

**Review by artonge — CHANGES_REQUESTED** [review]

File: lib/base.php (hunk 3 above)
Comment (mickenordin) [reply lib/base.php]: Fixed

File: lib/base.php (hunk 4 above)
Comment (mickenordin) [reply lib/base.php]: Fixed

**Review by artonge — APPROVED** [review]

**Review by Pytal — APPROVED** [review]

**Pull request merged**